Timetable and User business logic

Michal-MK opened the pull request.

Michal-MK committed:
Added raw Timetable service functionality

Michal-MK committed:
Added TimetableServiceTest class

Michal-MK committed:
Added UserService interface and implementation

Michal-MK committed:
Added UserServiceTest and added annotations to to the test classes to make running tests from IDEA easier

Michal-MK committed:
Added UserFacade and its implementation
- added RegisterUserDTO and UserContactInfoDTO
- still WIP

Michal-MK committed:
Added TimetableFacade and its implementation
- added TimetableDTO, AddTimetableDTO, TimetableEntryDTO and TimetableChatMessageDTO
- still WIP

Michal-MK committed:
Changed getTimetable to use mapService
- added getTimetableWithEntries

Michal-MK committed:
Replaced mock entities with the real deal

Michal-MK committed:
Added setCredits and changePhoneNumber to UserFacade

Michal-MK committed:
Added OfferPopularityService
- the actual complex logic in service layer
- added getSubscribedOffers to OfferDao for populating lazy-loaded list in User

Michal-MK committed:
Aaaa long commit
- added getByIdWithUser to eagerly fetch the user data to DAO and Service
- fixed getTimetable to properly map internal fields, for some reason it is not done automatically
- changed UserContactInfoDTO because the mapper sucks
- registerUser now returns the ID of the User
- added UserCreditsDTO
- reimplemented Facade tests in an "integration" style

Michal-MK committed:
Fixed TimetableEntry removal.
- when no entries exist an empty list is returned instead of an exception

JirkaVrbka commented on business/src/test/java/cz/idomatojde/UserServiceTest.java:
```diff
@@ -0,0 +1,90 @@
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
```
unused import

JirkaVrbka commented on business/src/test/java/cz/idomatojde/facade/TimetableFacadeTest.java:
```diff
@@ -0,0 +1,57 @@
+import cz.idomatojde.dto.timetable.AddTimetableDTO;
+import cz.idomatojde.dto.timetable.TimetableEntryDTO;
+import cz.idomatojde.dto.user.RegisterUserDTO;
+import cz.idomatojde.dto.timetable.TimetableDTO;
```
unused import

JirkaVrbka commented on business/src/test/java/cz/idomatojde/facade/TimetableFacadeTest.java:
```diff
@@ -0,0 +1,57 @@
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+
+import java.util.Calendar;
```
unused import

JirkaVrbka commented on dao/src/main/java/cz/idomatojde/dao/OfferDao.java:
```diff
@@ -20,6 +20,14 @@
     List<Offer> findByUser(User user);
 
+    /**
+     * Finds all offers a {@link User} is sibscribed to
```
Typo in "sibscribed"

JirkaVrbka commented on api/src/main/java/cz/idomatojde/facade/TimetableFacade.java:
```diff
@@ -0,0 +1,27 @@
+     * Create a new timetable
+     *
+     * @param timetableDTO necessary timetable creation information
+     * @return uniqe id of the timetable
```
typo in "uniqe"

JirkaVrbka commented on api/src/main/java/cz/idomatojde/facade/UserFacade.java:
```diff
@@ -0,0 +1,53 @@
+     * @param registrationInfo necessary user creation information
+     * @return the ID of the User
+     */
+    Long registerUser(RegisterUserDTO registrationInfo);
```
It is not necessary to return Long, add Id to the DTO (as reference project does)

Michal-MK replied:
IMO, the DTO should contain only the info necessary for registration, the User's ID is not necessary for a registration to succeed.

JirkaVrbka replied:
I would refer you to reference project. They have it like that

Michal-MK replied:
Are you sure? [Link](https://i.gyazo.com/3d240ae82045388a401c9c62df37a2cf.png)

JirkaVrbka replied:
[https://github.com/fi-muni/PA165/blob/master/eshop-service/src/main/java/cz/fi/muni/pa165/service/facade/UserFacadeImpl.java](link)

They have inconsistency as well. Let me change the code and return Id. Actually, I like it more

Michal-MK replied:
Interesting, I just continued with what we did in DAO and Service layers to be consistent.

JirkaVrbka commented on business/src/main/java/cz/idomatojde/services/TimetableServiceImpl.java:
```diff
@@ -0,0 +1,112 @@
+
+
+    @Override
+    public void updateEntry(TimetableEntry entry) {
```
Updating is done automatically, no need for this method

JirkaVrbka commented on business/src/main/java/cz/idomatojde/services/facades/TimetableFacadeImpl.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @Override
+    public long addTimetable(AddTimetableDTO timetableDTO) {
```
Same as comment before, add Id to DTO

JirkaVrbka commented on business/src/main/java/cz/idomatojde/services/facades/UserFacadeImpl.java:
```diff
@@ -0,0 +1,70 @@
+    public void changePhoneNumber(long userId, String phoneNumber) {
+        User u = userService.getById(userId);
+
+        // TODO Phone number verification
```
TODO?

JirkaVrbka commented on business/src/test/java/cz/idomatojde/TimetableServiceTest.java:
```diff
@@ -0,0 +1,364 @@
+    private final LocalTime TIME_4PM = LocalTime.of(16, 0);
+    private final Duration DUR_2H = Duration.ofHours(2);
+
+    private User DEF_USER;
```
Why upper cases?

JirkaVrbka commented on business/src/test/java/cz/idomatojde/UserServiceTest.java:
```diff
@@ -0,0 +1,90 @@
+
+    private UserService service;
+
+    private User DEF_USER;
```
Why upper cases?

Michal-MK committed:
Fixes
- fixed typos in JavaDoc
- removed unused imports
- removed unnecessary update functions
- fixed naming for non-const fields
- hastily implemented phone number validation regex, this may not be final
-
